tutorials/text/llama-nemotron-data-curation/filters/model_filters.py

```diff
@@ -0,0 +1,371 @@
+        scores = pd.Series([len(tokens) for tokens in tokenized["input_ids"]], index=df.index)
+        mask = (scores > 0) & (scores <= self.max_token_count)
+        df_filtered = df[mask]
+
+        return DocumentBatch(
+            task_id=batch.task_id,
+            dataset_name=batch.dataset_name,
+            data=df_filtered,
+            _metadata=batch._metadata,
+            _stage_perf=batch._stage_perf,
+        )
+
+
+# Tokenize text and filter out samples with too many tokens
+class CompletionTokenCountFilter(ProcessingStage[DocumentBatch, DocumentBatch]):
+    def __init__(
+        self,
+        tokenizer_identifier: str,
+        hf_token: str | None = None,
+        max_completion_token_count: int = 8192,
+        output_field: str = "output",
+    ):
+        self._name = "completion_token_count_filter"
+        self.tokenizer_identifier = tokenizer_identifier
+        self.hf_token = hf_token
+        self.max_completion_token_count = max_completion_token_count
+        self.output_field = output_field
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], [self.output_field]
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], []
+
+    def ray_stage_spec(self) -> dict[str, Any]:
+        return {"is_actor_stage": True}
+
+    def setup_on_node(self, _node_info: NodeInfo | None = None, _worker_metadata: WorkerMetadata = None) -> None:
+        try:
+            self._setup(local_files_only=False)
+        except Exception as e:
+            msg = f"Failed to download {self.tokenizer_identifier}"
+            raise RuntimeError(msg) from e
+
+    # We use the _setup function to ensure that everything needed for the tokenizer is downloaded and loaded properly
+    def _setup(self, local_files_only: bool = True) -> None:
+        self.tokenizer = AutoTokenizer.from_pretrained(
+            self.tokenizer_identifier,
+            local_files_only=local_files_only,
+            token=self.hf_token,
+        )
+
+    def setup(self, _: WorkerMetadata | None = None) -> None:
+        self._setup(local_files_only=True)
+
+    def process(self, batch: DocumentBatch) -> DocumentBatch:
+        df = batch.to_pandas()
+
+        if df.empty:
+            batch.data["completion_token_count"] = []
+            return batch
+
+        outpt = df[self.output_field]
+        outpt_copy = outpt.copy()
+
+        templates_list = outpt_copy.apply(
+            lambda text: self.tokenizer.apply_chat_template(
+                [{"role": "assistant", "content": text}],
+                # For maximum accuracy it should be tokenize=True here,
+                # but for speedups we just use the length of the base string instead of tokenizing it
+                # (we consider this to be acceptable since 1 token is approx 4 characters for English
+                # and we are only using the CompletionTokenCountFilter as a proxy for text complexity)
+                tokenize=False,
```

Comment:
IIRC we aligned on this when implementing the Dask tutorial. Just wanted to double check that this is still okay.